chronos/selectors2.nim

```diff
@@ -285,78 +281,45 @@ else:
+        else:
+          ok()
       else:
-        if posix.sigprocmask(SIG_BLOCK, newmask, oldmask) == -1:
-          raiseIOSelectorsError(osLastError())
+        if sigprocmask(SIG_BLOCK, newmask, oldmask) == -1:
+          err(osLastError())
+        else:
+          ok()
 
-    template unblockSignals(newmask: var Sigset, oldmask: var Sigset) =
+    proc unblockSignals(newmask: var Sigset,
+                        oldmask: var Sigset): Result[void, OSErrorCode] =
       when hasThreadSupport:
-        if posix.pthread_sigmask(SIG_UNBLOCK, newmask, oldmask) == -1:
-          raiseIOSelectorsError(osLastError())
+        if pthread_sigmask(SIG_UNBLOCK, newmask, oldmask) == -1:
+          err(osLastError())
+        else:
+          ok()
       else:
-        if posix.sigprocmask(SIG_UNBLOCK, newmask, oldmask) == -1:
-          raiseIOSelectorsError(osLastError())
+        if sigprocmask(SIG_UNBLOCK, newmask, oldmask) == -1:
+          err(osLastError())
+        else:
+          ok()
 
-  template clearKey[T](key: ptr SelectorKey[T]) =
-    var empty: T
-    key.ident = InvalidIdent
-    key.events = {}
-    key.data = empty
-
-  proc verifySelectParams(timeout: int) =
+  template verifySelectParams(timeout, min, max: int) =
     # Timeout of -1 means: wait forever
     # Anything higher is the time to wait in milliseconds.
-    doAssert(timeout >= -1, "Cannot select with a negative value, got " & $timeout)
-
-  when defined(linux):
-    include ./ioselects/ioselectors_epoll
-  elif bsdPlatform:
-    include ./ioselects/ioselectors_kqueue
-  elif defined(windows):
-    include ./ioselects/ioselectors_select
-  elif defined(solaris):
-    include ./ioselects/ioselectors_poll # need to replace it with event ports
-  elif defined(genode):
-    include ./ioselects/ioselectors_select # TODO: use the native VFS layer
-  elif defined(nintendoswitch):
-    include ./ioselects/ioselectors_select
-  else:
-    include ./ioselects/ioselectors_poll
-
-proc register*[T](s: Selector[T], fd: int | SocketHandle,
-                  events: set[Event], data: T) {.deprecated: "use registerHandle instead".} =
-  ## **Deprecated since v0.18.0:** Use ``registerHandle`` instead.
-  s.registerHandle(fd, events, data)
-
-proc setEvent*(ev: SelectEvent) {.deprecated: "use trigger instead",
-    raises: [Defect, IOSelectorsException].} =
-  ## Trigger event ``ev``.
-  ##
-  ## **Deprecated since v0.18.0:** Use ``trigger`` instead.
-  ev.trigger()
-
-proc update*[T](s: Selector[T], fd: int | SocketHandle,
-                events: set[Event]) {.deprecated: "use updateHandle instead".} =
-  ## Update file/socket descriptor ``fd``, registered in selector
-  ## ``s`` with new events set ``event``.
-  ##
-  ## **Deprecated since v0.18.0:** Use ``updateHandle`` instead.
-  s.updateHandle()
+    doAssert((timeout >= min) and (timeout <= max),
+             "Cannot select with incorrect timeout value, got " & $timeout)
+
+when asyncEventEngine == "epoll":
```

Comment:
shouldn't `iocp` be an engine also, which replaces `when windows`?

Reply:
It could be, but there no other engines available on Windows. For example linux could use `epoll` and `poll`, BSD/MacOS could also use `linux` and `poll`, but Windows support is different and already bundled inside asyncloop, it do not use selectors2 module.

Reply:
https://github.com/piscisaureus/wepoll :)

In general though, it feels a bit more clean if we use a single mechanism for selecting "backend" throughout the source code - it makes the code more easy to understand because there's only one "concept" involved - who knows, maybe future windows versions will support sockets in I/O ring or something similar as well?

I'd go for it, since we're here already - potentially it would then make sense to rename `ioselects` to `backends`

Reply:
`WSAPoll` and `select` (implementation which i have removed from the repository) support only sockets on Windows. Pipes and all other things are not supported. There is no reason to keep implementation which do not support all the `chronos` features. Even with `ioselectors_poll` and Posix system with some effort its possible to support all the chronos features (yeah we will need to create threads and all other things, but its possible). But whatever we going to do its impossible to achieve with `WsaPoll` and `select`. So i do not see any reasons to support this mechanisms.